### DISCUSSION TITLE
Will the number of axial nodes influence the convergence?

### DISCUSSION THREAD

**Discussion by Ethan-xj** [idaholab/moose#22615] [Q&A General] [answered]
When I set about 10 nodes in a 1D line , the case can converge, but when I add the nodes to 20 or more, the case can't converge. In my understanding, if there's no mistake in the equations, the convergence should be better with the increase of the nodes. So does this prove that there stiil exists some mistakes in my equations?

Here is the fourth equation's method:
I use the [CoupledVariableValueMaterial](https://mooseframework.inl.gov/modules/thermal_hydraulics/source/materials/CoupledVariableValueMaterial.html) to stores values of a variable into material properties and use [Sampler1DReal](https://mooseframework.inl.gov/modules/thermal_hydraulics/source/vectorpostprocessors/Sampler1DReal.html) to export the vector which contains the value at every qppoint of the variables. And I write a userobject to output the variable value based on the coordinate of the qppoint. At last I used the userobjectInterface to couple the userobject and the kernel to express the variable value of other block in the current block. By the way, the "execute on" in vectorpostprocessors is "INITIAL LINEAR NONLINEAR". I'm not sure if this method is correct mathematically. 
I'd like you to discuss it with me. Thank you.

Ethan

**Comment by GiudGiud** [comment, marked as answer]
Hello

When you plot the variable you manually transfered to the other block, does it look right?

Is there two-way coupling between these two blocks? Should you add contributions to the Jacobian from this coupling?

This is our current page for debugging these issues
https://mooseframework.inl.gov/application_usage/failed_solves.html
I ll add that you can try using svd to examine the condition number

Guillaume

**Reply by Ethan-xj** [reply]
Yes, two-way coupling. I used the method I mentioned to solve the equation which contain other block's variable value (like we've discussed before:https://github.com/idaholab/moose/discussions/22425#discussioncomment-3943351 ). And I used AD in all kernels. So I don't think I need to do anything to the Jacobian.
As for the condition number, the situation is like:
```
SVD: condition number 5.589344305504e+11, 0 of 148 singular values are (nearly) zero
      SVD: smallest singular values: 9.035132051683e-06 1.777040744215e-05 2.356642901462e-05 3.278996559336e-05 3.849557833974e-05
      SVD: largest singular values : 4.648359214010e+06 4.850303332068e+06 4.911589108063e+06 5.043440409533e+06 5.050046388255e+06
```
It's large but can converge.
I have browsed the page you sent before. And just now I noticed that when I add the number of axial nodes, the residual acts like linear did not converge. Maybe I should see the Scaling and seek inspiration. By the way, do you think the approach I used is correct mathematically?

**Reply by GiudGiud** [reply]
You used AD but the transfer scheme you used did not preserve the derivatives
First you need to use the ADCoupledValueFunctionMaterial not the CoupledValueFunctionMaterial
but then there is not version of the Sampler that exports derivatives. 

You ll have to fill the Jacobian manually.
You can look at the quality of your current Jacobian with this page
https://mooseframework.inl.gov/help/development/analyze_jacobian.html

**Reply by Ethan-xj** [reply]
So that's it. Thank you! I will check my Jacobian first.

**Reply by GiudGiud** [reply]
btw if your Jacobian isnt very good, then we recommend using PJFNK over Newton for the solver method. 
That could be an alternative

**Reply by Ethan-xj** [reply]
```
Kernel for variable 'p':
  (0,1) Off-diagonal Jacobian for variable 'u' is slightly off (by 0.022705 %)

Kernel for variable 'T':
  (2,2) On-diagonal Jacobian is slightly off (by 0.421074 %)

Kernel for variable 'w':
  (3,0) Off-diagonal Jacobian for variable 'p' is wrong (off by 29.3 %)
  (3,2) Off-diagonal Jacobian for variable 'T' is wrong (off by 29.3 %)
```
The Jacobian has problem indeed. So the Jacobian Debugger won't tell me how to modify? I need to fix it by myself, right?

**Reply by Ethan-xj** [reply]
And how should I fix the Jacobian when I use AD?

**Reply by GiudGiud** [reply]
We should take a step back and consider what are the options here.
Are you 100% sure you want to do Newton here?
A fixed point (or quasi-Newton, available as well) scheme with multiapps or PJFNK with your current setup could be good enough?

If you want to do Newton, there will only be two options:
- You can modify derivatives of AD numbers manually when computing the residual.
- You write a user object that performs the transfer, including the derivatives, and either makes it directly available in the kernel or is used by a material property (auxiliary variables dont support AD)

**Reply by Ethan-xj** [reply]
Wait I think I might have some misunderstanding about the numerical algorithm in MOOSE(as far as I know, Newton and PJFNK). I always use the PJFNK in my input file and I discuss the Jacobian with you is because I always think that PJFNK need the Jacobian. So if I use AD in kernel and PJFNK in excutioner at the same time, how does MOOSE calculate?
The Jacobian debugger can only verify the correctness of the Newton?
And as far as I have already found the mistake in my Jacobian, if I use PJFNK, can I avoid this mistake? Or I still need to do something to fix the mistake? Maybe I should not use AD and write the Jacobian all by myself.

**Reply by GiudGiud** [reply]
With AD and PJFNK, AD is only used to help with building the preconditioner, which sometimes involves building the Jacobian

AD is mostly helpful for Newton's method for building good Jacobians at every step.

**Reply by Ethan-xj** [reply]
OK I got it. I've read the [Preconditioning page](https://mooseframework.inl.gov/syntax/Preconditioning/index.html) again and I don't understand the relationship between (RT)T and (k(s,T)▽ϕj,▽ψj). The preconditioning matrix is built by MOOSE and I can't make sure that the preconditioning matrix is correct. So if I use Newton, I need to modify the Jacobian manually. If I use PJFNK, I still need to know whether the preconditioning matrix is correct or not. And if the matrix is not correct, how can I modify the preconditioning matrix?

**Reply by GiudGiud** [reply]
It depends on the kind of preconditioner you use. Some will only use the diagonal of the Jacobian for example.

**Reply by GiudGiud** [reply]
either way to improve the Jacobian or the preconditioning, you ll likely want to code part of the Jacobian manually or find a way to transfer the AD numbers.

**Reply by Ethan-xj** [reply]
Can I ask what is the AD number? I haven't heard it before.

**Reply by GiudGiud** [reply]
Here the AD numbers you want to transfer are the variable values in the other part of the domain. 

How complex is the mapping from one block to another? What information do you need in a given element to know which element from the other block you need to get values for? is it just a translation?

**Reply by Ethan-xj** [reply]
OK. I'm solving the lateral flow between two 1D flow channels. So I need to express the physical parameter(average value of two channels) of the gap. I think the mapping isn't complex, I give the two channels same axial nodes setting and use the coordinate of qppoint as the bridge. At every linear and nonlinear I get two vectors(other block's variable value and coordinate of qppoint) contained by vectorpostprocessors. And in the current kernel, the userobject use the current coordinate of qppoint to search the "coordinate of qppoint vector" and output the corresponding variable value.

**Reply by Ethan-xj** [reply]
I'm not sure if it's mapping. I think it's a kind of manual-mapping XD.

**Reply by GiudGiud** [reply]
oh I didnt realize this was 1D
I think this has already been done for some THM applications
@joshuahansel

**Reply by Ethan-xj** [reply]
you mean the value translation between different blocks？Can you tell me which THM app it is？

**Reply by joshuahansel** [reply]
In THM we can couple a 1D channel to the side of a 2D heat structure. See the `HeatTransferFromHeatStructure1Phase` component. We get all of the derivatives through AD. We have not tried coupling two 1D channels together, but it should be the same principle. The general strategy would be this:

- Transfer the relevant quantities from flow channel 2 to flow channel 1 using something like `VariableValueTransferMaterial`
- Execute an elemental UO like `ADHeatFluxFromHeatStructure3EqnUserObject` that computes and caches the coupling terms.
- Have your kernels (like `ADOneD3EqnEnergyHeatFlux`) take the desired coupling terms from the UO.

There is a class called `FlowChannelAlignment` that builds the mapping between the 1D channel and 2D heat structure, which is used by the UO. You could make something similar for 1D to 1D mapping.

**Reply by Ethan-xj** [reply]
Thank you!

**Reply by Ethan-xj** [reply]
@GiudGiud  Hey, I want to ask that is the JacobianDebugger reliable? There is a derivative term of the variable 'u' in a equation which main variable is 'w'. And the debugger says: 
```
Kernel for variable 'w':
 (3,1) Off-diagonal Jacobian for variable 'u' should just return  zero
```
I don't know why.

**Reply by GiudGiud** [reply]
Yes it s supposed to be. 

The term in question is added properly to the residual and the jacobian? Can you share the code for the kernel?

**Reply by Ethan-xj** [reply]
This is the residual for 'w'. _lvel stands for 'w' and _vel stands for 'u':
```
Real
lateralmomentum_newton::computeQpResidual()
{
  Real qpx = _q_point[_qp](0);
  Real _tempadj = _qpvalueTEMP.computeQpValue(qpx);
  Real _rhoadj;
  double h = _current_elem->hmin();
  double deltax = h/10000;
  if (_tempadj<= 553.75 && _tempadj>=273.15)
     {
        _rhoadj = 838.466135+1.40050603*_tempadj-0.0030112376*pow(_tempadj,2)+3.71822313e-7*pow(_tempadj,3);
     }
  else
     {
       _rhoadj = 838.466135+1.40050603*553.75-0.0030112376*pow(553.75,2)+3.71822313e-7*pow(553.75,3);
     }
  Real _pressadj = _qpvaluePRESS.computeQpValue(qpx);
  Real _veladj = _qpvalueVEL.computeQpValue(qpx);
  Real _veladjdot = _linearVEL.computeQpValue(qpx+deltax);
//  ADReal _veladjdot = _linearVEL.computeQpValue(qpx+deltax);
  Real normal_part = _test[_i][_qp]*(_is_same_direction)*(1/2*(_lvel[_qp]*_grad_vel[_qp](_1Ddirection)
                     +_lvel[_qp]*(_veladjdot-_veladj)/deltax
                    +_grad_lvel[_qp](_1Ddirection)*(_vel[_qp]+_veladj))-0.5*(_press[_qp]-_pressadj)+
                    _kg*fabs(_lvel[_qp])*_lvel[_qp]/((_rho[_qp]+_rhoadj)*_sk*_lk));
  return  normal_part;
}
```
This is the Jacobian:
```
Real
lateralmomentum_newton::computeQpJacobian()
{
  Real qpx = _q_point[_qp](0);
  Real _tempadj = _qpvalueTEMP.computeQpValue(qpx);
  Real _rhoadj;
  double h = _current_elem->hmin();
  double deltax = h/10000;
  if (_tempadj<= 553.75 && _tempadj>=273.15)
     {
        _rhoadj = 838.466135+1.40050603*_tempadj-0.0030112376*pow(_tempadj,2)+3.71822313e-7*pow(_tempadj,3);
     }
  else
     {
       _rhoadj = 838.466135+1.40050603*553.75-0.0030112376*pow(553.75,2)+3.71822313e-7*pow(553.75,3);
     }
  Real _pressadj = _qpvaluePRESS.computeQpValue(qpx);
  Real _veladj = _qpvalueVEL.computeQpValue(53.75 && _tempadj>=273.15)
     {
        _rhoadj = 838.466135+1.40050603*_tempadj-0.0030112376*pow(_tempadj,2)+3.71822313e-7*pow(_tempadj,3);
     }
  else
     {
       _rhoadj = 838.466135+1.40050603*qpx);
  Real _veladjdot = _linearVEL.computeQpValue(qpx+deltax);
  Real normal_part = _test[_i][_qp]*(_is_same_direction)*(1/2*(_phi[_j][_qp]*_grad_vel[_qp](_1Ddirection)
                     +_phi[_j][_qp]*(_veladjdot-_veladj)/deltax
                    +_grad_phi[_j][_qp](_1Ddirection)*(_vel[_qp]+_veladj))+_kg*(fabs(_u[_qp])*_phi[_j][_qp]+fabs(_phi[_j][_qp])*_u[_qp])/((_rho[_qp]+_rhoadj)*_sk*_lk));
  return normal_part;
}
```
This is the offJacobian:
```
Real
lateralmomentum_newton::computeQpOffDiagJacobian(unsigned jvar)
{
  if(jvar == _vel_var_number)
  {
    Real qpx = _q_point[_qp](0);
    double h = _current_elem->hmin();
    double deltax = h/10000;
    Real _veladj = _qpvalueVEL.computeQpValue(qpx);
    Real _veladjdot = _linearVEL.computeQpValue(qpx+deltax);
    Real _lat_kvel = (_veladjdot-_veladj)/deltax;
    Real normal_part = _test[_i][_qp]*(_is_same_direction)*1/2*(_u[_qp]*_grad_phi[_j][_qp](_1Ddirection)+_grad_u[_qp](_1Ddirection)*(_phi[_j][_qp]+_lat_kvel));
    return normal_part;
  }
  else if(jvar == _press_var_number)
  {
    Real qpx = _q_point[_qp](0);
    double h = _current_elem->hmin();
    double deltax = h/10000;
    Real _pressadj = _qpvaluePRESS.computeQpValue(qpx);
    Real _pressadjdot = _linearPRESS.computeQpValue(qpx+deltax);
    Real _lat_kpress = (_pressadjdot-_pressadj)/deltax;
    Real normal_part = -0.5*_test[_i][_qp]*(_is_same_direction)*(_phi[_j][_qp]-_lat_kpress);
    return normal_part;
  }
  else
  {
    return 0.0;
  }
}

```
By the way, if the material property is a function of the variable, should I take the derivative of it in Jacobian? like density, function of temperature.

**Reply by GiudGiud** [reply]
What is this computeQpValue routine? What quantities are _qpvalueVEL?

Why are you doing finite difference gradients? What are these objects?

For the material property derivative, it depends on the equation and its Jacobian.
If a derivative showed up when you derived the Jacobian, you will likely need to have it in the contribution to the Jacobian

You can check the Jacobian with this tool
https://mooseframework.inl.gov/help/development/analyze_jacobian.html

**Reply by Ethan-xj** [reply]
What is this computeQpValue routine?
computeQpValue is a UO function which achieve one-to-one mapping between two vectors(qp value and qp coordinate). So computeQpValue(qpx) is the right value I want when qpx is the current qp coordinate.
What quantities are _qpvalueVEL?
`_qpvalueVEL(getUserObject<qpvaluebyx>("qpvalueVEL"))`
computeQpValue is defined in qpvaluebyx.C

I don't know about the mesh mapping before. So I can only transfer the value between two blocks, I can only write the derivative by myself. And I'm trying to add the contribution of the value from UO to the Jacobian manually, But it's hard. As you say, I should have the material paoperty in the contribution to the Jacobian, it's a trouble.

**Reply by GiudGiud** [reply]
@joshuahansel Ethan is trying to couple to non-local DOFs for 2 adjacent 1D systems. Is there something in THM they can look at for this?
They're going down a path of writing the Jacobian contributions manually, but it's getting complicated with the nonlocal dofs

**Reply by Ethan-xj** [reply]
@GiudGiud @joshuahansel  joshuahansel has advised me before and I'm focused on the mapping between two 1D mesh, that's why I ask you the question about the KDTree which is uesd in the THM block.
Thank you both very much!

**Reply by joshuahansel** [reply]
Ok, from what I gather, I'm guessing that you were just trying to test the derivatives that you have already implemented manually to verify that you do indeed have Jacobian issues. I believe that you're trying to work through the AD approach I outlined. Just let me know if you have questions on that.

**Reply by Ethan-xj** [reply]
@joshuahansel The AD approach is similar with the approach I used to use except the way to transfer other block's value, so I have no question about the approach.
But I have some doubt about the mapping code indeed:
[FlowChannelAlignment.C](https://github.com/idaholab/moose/blob/3e5d9cd59e32dcde8ed651000c8ee5f9520fa35d/modules/thermal_hydraulics/src/utils/FlowChannelAlignment.C)
```
  if (_master_points.size() > 0 && _slave_points.size() > 0)
  {
    // find the master elements that are nearest to the slave elements
    KDTree kd_tree(_master_points, _mesh.getMaxLeafSize());
    for (std::size_t i = 0; i < _slave_points.size(); i++)
    {
      unsigned int patch_size = 1;
      std::vector<std::size_t> return_index(patch_size);
      kd_tree.neighborSearch(_slave_points[i], patch_size, return_index);

      _nearest_elem_ids.insert(
          std::pair<dof_id_type, dof_id_type>(slave_elem_ids[i], master_elem_ids[return_index[0]]));
      _nearest_elem_ids.insert(
          std::pair<dof_id_type, dof_id_type>(master_elem_ids[return_index[0]], slave_elem_ids[i]));
    }
  }

```
I think the size of the vector `return_index` is the number of the nearest nodes of the query point `_slave_points[i]`.
And I want to know what's the value of `return_index[0]`?
`master_elem_ids[return_index[0]]` should be the index of the nearest node right?

Ethan

**Reply by joshuahansel** [reply]
By the way, just to clear up some language: "master" here refers to the heat structure side, and "slave" refers to the 1D flow channel. I plan to clean up this code at some point to use clearer language...

> I think the size of the vector return_index is the number of the nearest nodes of the query point _slave_points[i].

Yes, and it looks like it's just fixed to be size-1, so just a single match.

> And I want to know what's the value of return_index[0]?

Yes, here `return_index[0]` should be the vector index (in `_master_points`) corresponding to the nearest neighbor to `_slave_points[i]`. So then `master_elem_ids[return_index[0]]` should be the element ID of the nearest heat structure element, since the ordering in `master_elem_ids` and `_master_points` is the same.

**Reply by Ethan-xj** [reply]
@joshuahansel Thank you! And I have some questions about the input file. I know the master refers to the 2D element side and slave refers to the 1D element, but I don't know how to define this relationship in input file. For example, in [ADHeatFluxFromHeatStructure3EqnUserObject](https://mooseframework.inl.gov/modules/thermal_hydraulics/source/userobjects/ADHeatFluxFromHeatStructure3EqnUserObject.html), what should I fill in the `_fch_alignment`? I can't find any examples to refer to.

Ethan

**Reply by joshuahansel** [reply]
That's right, we never build that UO from an input file, only from within a `Component`. That parameter `_fch_alignment` is a pointer parameter, and there's no way to set that in an input file. You'll probably need to create your own component like `HeatTransferFromHeatStructure1Phase`. I'd suggest that you can build that mapping in the UO itself, but I'm not sure that's even possible, since you need to augment the sparsity pattern with the mapping, as in `HeatTransferFromHeatStructure1Phase::setupMesh()`. And it'd definitely be a lot more work.

**Reply by Ethan-xj** [reply]
@joshuahansel I got it. But I have the feeling that the problem is more and more complicated. Anyway, I will have a try. Thank you!

**Reply by Ethan-xj** [reply]
@joshuahansel I check the `HeatTransferFromHeatStructure1Phase` and I find that `Component` is significant in THM block. The relationship which I ever asked is achieved in `setupMesh()` using `Component`. I think it's hard for me to create my own component since there's lots of program skills in the code which I can't understand.
And I want to know why should I augment the sparsity pattern if I build the mapping in the UO? I haven't heard the sparsity pattern before.

**Reply by joshuahansel** [reply]
Yes, I think what you're trying to do is not easy for someone who is not an experienced THM developer. I think that a general channel-to-channel coupling is possibly a useful capability in THM. I'd like to do it myself, but I can't work on this right now. What physical coupling are you trying to achieve? For example, are the channels just exchanging heat or what? If just exchanging heat, I'd suggest just having a heat structure between the channels to exchange the heat. Of course, if you have some other kind of coupling, then nevermind - you'll need what you're trying to do now.

The sparsity pattern is which entries in the Jacobian matrix are nonzero. It's usually determined automatically based on the size of the support of the finite elements, but here since we're coupling subdomains together ourselves, we need to modify the sparsity pattern ourselves.

**Reply by Ethan-xj** [reply]
Thanks!
The channels will exchange energy and momentum so the there will be a lateral momentum equation to solve the lateral velocity and there will be a cross-flow term added in the energy and momentum equation. I think twice and decide to develop the code based on the THM, especially the `Component` block. In my understanding, the variables in some components, like `FlowChannel1Phase` are defined and fixed. So if I wanna add some kernels or some terms in kernel, I must use the current variable names right?
By the way, THM is complicated indeed. I'm sure I will have more questions about THM. 

Ethan

**Reply by joshuahansel** [reply]
That's right, the variable names are fixed.

**Reply by Ethan-xj** [reply]
@joshuahansel I write the channel-to-channel mapping with reference to `FlowChannelAlignment`. You know, I delete the content about the element side and the mapping between master element side and master element id. Now the mapping only exists in the master element id and slave element id. But I have a question: How can I test this mapping? It will be too troublesome to write all the code I need include the new 2channels component. Do you have any good ideas?

Ethan

**Reply by joshuahansel** [reply]
So you want to test your new mapping, but without writing your new component? You could create a simplified component for the purpose of testing, but I don't think it would be too much less than writing your component. Personally, I wouldn't be worried that the mapping is correct, since it should be easy enough to fix. I would assume it works and then write your component and test that.

**Reply by Ethan-xj** [reply]
Okay, I think so.
By the way, if I want to add a new nonlinear variable in the `FlowChannel1Phase` component, should I just add the variable in the FlowModelSinglePhase.C and FlowModelSinglePhase.h just like RHOA RHOUA and RHOEA?
I have tried to understand the logic but I'm not sure. You know, the `FlowChannel1Phase` has many base document.

**Reply by joshuahansel** [reply]
Yes, you'll inherit from `FlowModelSinglePhase`, override `addVariables()` with something like
```
void
MyNewFlowModel::addVariables()
{
  FlowModelSinglePhase::addVariables();

  const std::vector<SubdomainName> & subdomains = _flow_channel.getSubdomainNames();

  _sim.addSimVariable(true, "my_new_var", _fe_type, subdomains, scaling_factor_for_my_new_var);

  _solution_vars = {RHOA, RHOUA, RHOEA, "my_new_var"};
  _derivative_vars = _solution_vars;
}
```

**Reply by Ethan-xj** [reply]
It really helped a lot! Thank you!
